examples/graph/gat_node_classification.py

```diff
@@ -0,0 +1,398 @@
+        attention_scores_sum = tf.math.unsorted_segment_sum(
+            data=attention_scores,
+            segment_ids=edges[:, 0],
+            num_segments=tf.reduce_max(edges[:, 0]) + 1,
+        )
+        attention_scores_sum = tf.repeat(
+            attention_scores_sum, tf.math.bincount(tf.cast(edges[:, 0], "int32"))
+        )
+        attention_scores_norm = attention_scores / attention_scores_sum
+
+        # (4) Gather node states of neighbors, apply attention scores and aggregate
+        node_features_neighbors = tf.gather(node_features_transformed, edges[:, 1])
+        out = tf.math.unsorted_segment_sum(
+            data=node_features_neighbors * attention_scores_norm[:, tf.newaxis],
+            segment_ids=edges[:, 0],
+            num_segments=tf.shape(node_features)[0],
+        )
+        return out
+
+
+class MultiHeadGraphAttention(layers.Layer):
+    def __init__(self, units, num_heads=8, merge_type="concat", **kwargs):
+        super().__init__(**kwargs)
+        self.num_heads = num_heads
+        self.merge_type = merge_type
+        self.attention_layers = [GraphAttention(units) for _ in range(num_heads)]
+
+    def call(self, inputs):
+        atom_features, pair_indices = inputs
+
+        # Obtain outputs from each attention head
+        outputs = [
+            attention_layer([atom_features, pair_indices])
+            for attention_layer in self.attention_layers
+        ]
+        # Concatenate or average the node states from each head
+        if self.merge_type == "concat":
+            outputs = tf.concat(outputs, axis=-1)
+        else:
+            outputs = tf.reduce_mean(tf.stack(outputs, axis=-1), axis=-1)
+        # Activate and return node states
+        return tf.nn.relu(outputs)
+
+
+"""
+### Define model with custom `train_step`, `test_step` and `predict_step`
+"""
+
+
+class GraphAttentionNetwork(keras.Model):
+    def __init__(
+        self,
+        node_features,
+        adjacency,
+        hidden_units,
+        num_heads,
+        num_layers,
+        output_dim,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.node_features = node_features
+        self.adjacency = adjacency
+        self.preprocess = layers.Dense(hidden_units * num_heads, activation="relu")
+        self.attention_layers = [
+            MultiHeadGraphAttention(hidden_units, num_heads) for _ in range(num_layers)
+        ]
+        self.output_layer = layers.Dense(output_dim)
+
+    def call(self, inputs):
+        node_features, adjacency = inputs
+        x = self.preprocess(node_features)
+        for attention_layer in self.attention_layers:
+            x = attention_layer([x, adjacency]) + x
+        outputs = self.output_layer(x)
+        return outputs
+
+    def train_step(self, data):
+
+        indices, labels = data
+
+        with tf.GradientTape() as tape:
+
+            outputs = self([self.node_features, self.adjacency])
```

Comment:
So the model is data dependent? The overall mechanic at works deserves an explanation paragraph under `### Define model with custom `train_step`, `test_step` and `predict_step``

Reply:
The entire graph (namely, `self.node_features` and `self.adjacency`) is operated on both during training and testing. However, the loss is only calculated for the training nodes (i.e., `loss(train_labels, tf.gather(gat_output, train_indices))`). During testing we instead gather the `test_indices`. I could try to add some information regarding this.

Reply:
The same is done [in this tutorial](https://keras.io/examples/graph/gnn_citations/#implement-a-graph-neural-network-node-classifier). See `graph_info`.

Reply:
If you feel strongly about it, feel free to write a training loop from scratch without subclassing `Model`, up to you!

Reply:
Text/clarification has now been added to the subsection. I kept it as it was (subclassing `Model` with custom train/predict/test steps).